### PR TITLE
Perf: Add extended template version of timed_perf + use in CL perf

### PR DESCRIPTION
**Pull request by elcallio**
Adds sub-template for time_parallel with templated result type + optional per-iteration post-process func. Idea is that Res may be a subtype of perf_result, with additional stats, initiated on init, and post-process  function can fix up and apply stats -> we can add stats to result.

Then uses this mighty construct to add some IO stats to CL perf.

### CODE REVIEW

File: test/perf/perf.hh [hunk 1]
```diff
@@ -152,36 +151,45 @@ std::ostream& operator<<(std::ostream& os, const perf_result& result);
  *
  * Returns a vector of throughputs achieved in each iteration.
  */
-template <typename Func>
+template <typename Res, typename Func, typename UpdateFunc = void(*)(const Res&, const executor_shard_stats&)>
```
Comment (avikivity) [inline test/perf/perf.hh:154]: Let's add a constraint as a courtesy to the next user.

Comment (avikivity) [reply test/perf/perf.hh:154]: e.g. perf_result must be a base class, and std::invocable.

File: test/perf/perf.hh [hunk 2]
```diff
@@ -152,36 +151,45 @@ std::ostream& operator<<(std::ostream& os, const perf_result& result);
...
-            .mallocs_per_op = double(stats.allocations) / stats.invocations,
-            .tasks_per_op = double(stats.tasks_executed) / stats.invocations,
-            .instructions_per_op = double(stats.instructions_retired) / stats.invocations,
-        };
-        std::cout << result << "\n";
+
+        result.throughput = static_cast<double>(stats.invocations) / duration;
+        result.mallocs_per_op = double(stats.allocations) / stats.invocations;
+        result.tasks_per_op = double(stats.tasks_executed) / stats.invocations;
+        result.instructions_per_op = double(stats.instructions_retired) / stats.invocations;
```
Comment (avikivity) [inline test/perf/perf.hh:176]: my beautiful designated initializers

File: test/perf/perf_commitlog.cc [hunk 3]
```diff
@@ -42,6 +43,27 @@ struct test_config {
...
 };
 
+struct clperf_result : public perf_result {
+    double aio_writes;
+    double aio_write_bytes;
+
+    clperf_result()
+        : aio_writes(engine().get_io_stats().aio_writes)
+        , aio_write_bytes(engine().get_io_stats().aio_write_bytes)
+    {}
```
Comment (avikivity) [inline test/perf/perf_commitlog.cc:53]: We can add these to the standard thing, it's generic enough.

Comment (elcallio) [reply test/perf/perf_commitlog.cc:53]: You could, but then the next user comes and thinks "I wanted aio_reads as well..."
Better approach might be to add them as a shared sub-type in perf.hh (maybe via mix-in)

Comment (elcallio) [reply test/perf/perf_commitlog.cc]: Made it a mixin + extended shared type. Can be used by any bench if needed. CL now just type-defs + fixes its json output.

**Pull request closed without merging**